mongoose/mongoose.d.ts

```diff
@@ -269,14 +273,42 @@ declare module "mongoose" {
     path(path: string, constructor: any): Schema;
     pathType(path: string): string;
     plugin(plugin: (schema: Schema, options?: Object) => void, options?: Object): Schema;
-    post(method: string, fn: Function): Schema;
-    pre(method: string, callback: Function): Schema;
+
+    pre(method: string, fn: HookSyncCallback): Schema;
+    pre(method: string, fn: HookSyncCallback, errorCb: HookErrorCallback): Schema;
+    pre(method: string, isAsync: boolean, fn: HookAsyncCallback): Schema;
+    pre(method: string, isAsync: boolean, fn: HookAsyncCallback, errorCb: HookErrorCallback): Schema;
+    post(method: string, fn: HookSyncCallback): Schema;
+    post(method: string, fn: HookSyncCallback, errorCb: HookErrorCallback): Schema;
+    post(method: string, isAsync: boolean, fn: HookAsyncCallback): Schema;
+    post(method: string, isAsync: boolean, fn: HookAsyncCallback, errorCb: HookErrorCallback): Schema;
+
     requiredPaths(): string[];
     set(key: string, value: any): void;
     static(name: string, fn: Function): Schema;
     virtual(name: string, options?: Object): VirtualType;
     virtualpath(name: string): VirtualType;
+
+  }
+
+  // hook functions: https://github.com/vkarpov15/hooks-fixed
+  export interface HookSyncCallback {
+      (next: Function, ...hookArgs:any[]): any;
   }
+
+  export interface HookAsyncCallback {
+      (next: Function, done: Function, ...hookArgs:any[]): any;
+  }
+
+  export interface HookErrorCallback {
+      (error: Error): any;
+  }
+
+  export interface HookNextFunction {
```

Comment:
They are used in `pre(method: string, fn: HookSyncCallback, errorCb?: HookErrorCallback): Schema;`